pkgs/applications/version-management/sublime-merge/common.nix

```diff
@@ -0,0 +1,107 @@
+{ buildVersion, sha256, dev ? false }:
+
+{ fetchurl, stdenv, xorg, glib, gtk3, cairo, pango, libredirect, makeWrapper, wrapGAppsHook
+, pkexecPath ? "/run/wrappers/bin/pkexec", gksuSupport ? false, gksu
+, writeScript, common-updater-scripts, curl, gnugrep
+}:
+
+assert gksuSupport -> gksu != null;
+
+let
+  libPath = stdenv.lib.makeLibraryPath [xorg.libX11 glib gtk3 cairo pango];
+  redirects = [ "/usr/bin/pkexec=${pkexecPath}" ]
+    ++ stdenv.lib.optional gksuSupport "/usr/bin/gksudo=${gksu}/bin/gksudo";
+in let
+  # package with just the binaries
+  sublime_merge = stdenv.mkDerivation {
+    name = "sublime-merge-${buildVersion}-bin";
+    version = buildVersion;
+
+    src = fetchurl {
+      name = "sublime-merge-${buildVersion}.tar.xz";
+      url = "https://download.sublimetext.com/sublime_merge_build_${buildVersion}_x64.tar.xz";
+      inherit sha256;
+    };
+
+    dontStrip = true;
+    dontPatchELF = true;
+    buildInputs = [ glib gtk3 ]; # for GSETTINGS_SCHEMAS_PATH
+    nativeBuildInputs = [ makeWrapper wrapGAppsHook ];
+
+    buildPhase = ''
+      runHook preBuild
+
+      for binary in sublime_merge crash_reporter git-credential-sublime ssh-askpass-sublime; do
+        patchelf \
+          --interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" \
+          --set-rpath ${libPath}:${stdenv.cc.cc.lib}/lib64 \
+          $binary
+      done
+
+      # Rewrite pkexec|gksudo argument. Note that we can't delete bytes in binary.
+      sed -i -e 's,/bin/cp\x00,cp\x00\x00\x00\x00\x00\x00,g' sublime_merge
+
+      runHook postBuild
+    '';
+
+    installPhase = ''
+      runHook preInstall
+
+      # Correct sublime_merge.desktop to exec 'sublime_merge' instead of /opt/sublime_merge
+      substituteInPlace "sublime_merge.desktop" --replace "/opt/sublime_merge/sublime_merge" "$out/sublime_merge"
+
+      mkdir -p $out
+      cp -r * $out/
+
+      runHook postInstall
+    '';
+
+    dontWrapGApps = true; # non-standard location, need to wrap the executables manually
+
+    postFixup = ''
+      wrapProgram $out/sublime_merge \
+        --set LD_PRELOAD "${libredirect}/lib/libredirect.so" \
+        --set NIX_REDIRECTS ${builtins.concatStringsSep ":" redirects} \
+        "''${gappsWrapperArgs[@]}"
+    '';
+  };
+in stdenv.mkDerivation {
+  name = "sublime-merge-${buildVersion}";
+  version = buildVersion;
+
+  phases = [ "installPhase" ];
+
+  inherit sublime_merge;
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  installPhase = ''
+    mkdir -p "$out/bin"
+    makeWrapper "$sublime_merge/sublime_merge" "$out/bin/sublime_merge";
+    ln -s "$out/bin/sublime_merge" "$out/bin/smerge";
+    mkdir -p "$out/share/applications"
+    substitute "$sublime_merge/sublime_merge.desktop" "$out/share/applications/sublime_merge.desktop" --replace "$sublime_merge/sublime_merge" "$out/bin/sublime_merge"
+    for iconSize in $(ls "$sublime_merge/Icon"); do
```

Comment:
I tend to not use `ls` in scripts ever.

Using substring removal and a glob would be better.
```
for directory in $sublime_merge/Icon/*; do
  size=''${directory##*/}
  mkdir -p "$out/share/icons/hicolor/$size/apps";
  cp "$sublime_merge/Icon/$size/sublime-merge.png" "$out/share/icons/hicolor/$size/apps";
done
```

##### we need the `''` to escape the `$` btw

Reply:
Thanks for the feedback, I've fixed this.

Can you help me understand why it is good to avoid using ls? Doesn't it have reliable behavior, especially in a consistent environment like stdenv?

By the way I ended up using `$(dirname $directory)` instead of `''${directory##*/}` because I am more familiar with it. Could you point me to some documentation where I can learn more about the `##*/` shell construct? I'm having trouble finding anything that describes how that works.

Reply:
> By the way I ended up using $(dirname $directory) instead of ''${directory##*/} because I am more familiar with it. Could you point me to some documentation where I can learn more about the ##*/ shell construct? I'm having trouble finding anything that describes how that works.

This should be very helpful https://wiki-dev.bash-hackers.org/syntax/pe#substring_removal

> Can you help me understand why it is good to avoid using ls? Doesn't it have reliable behavior, especially in a consistent environment like stdenv?

The problem is that the output of `ls` is meant to be human readable and so its not meant to parsed.
We have globbing for this purpose, to allow the shell to expand patterns into a list of matching filenames.

Here's a useful resource on that topic
* http://mywiki.wooledge.org/ParsingLs

Reply:
That makes sense, thanks for the links!